### PR TITLE
Fix unorderedMatches for poorly ordered input

### LINKED ISSUE #73: `unorderedMatches` misses cases depending on the ordering of the input
```dart
// passes
expect([1, 2], unorderedMatches([greaterThan(0), greaterThan(1)]));

// fails
expect([2, 1], unorderedMatches([greaterThan(0), greaterThan(1)]));
```

### PR DESCRIPTION
**Pull request by natebosch**
Fixes #73

Reimplements the unordered matcher using a recursive search assuming
that matchers can match multiple values in the input rather than a
greedy algorithm which allows a matcher to "consume" a value that is
needed for some other matcher.

User visible differences:
- May match inputs that would have previously been (incorrectly)
  rejected.
- The failure description no longer includes the index of the matcher
  which is unmatched, and all unmatched matchers are printed rather than
  the first.

### CODE REVIEW

File: lib/src/iterable_matchers.dart [hunk 1]
```diff
@@ -206,6 +215,26 @@ class _UnorderedMatches extends Matcher {
   Description describeMismatch(item, Description mismatchDescription,
           Map matchState, bool verbose) =>
       mismatchDescription.add(_test(item));
+
+  /// Returns [true] if the value at [valueIndex] can be paired with some
+  /// unmatched matcher.
+  ///
+  /// Recursively looks for new pairings whenever there is a conflict. [seen]
+  /// tracks the matchers that have already been consumed within this search.
+  bool _findPairing(List<List<bool>> adjacency, int valueIndex, Set<int> seen,
```
Comment (jakemac53) [inline lib/src/iterable_matchers.dart]: nit: consider making `seen` optional and then doing `seen ??= new Set<int>();` so you don't have to pass in a new one when you call this method originally

File: lib/src/iterable_matchers.dart [hunk 2]
```diff
@@ -206,6 +215,26 @@ class _UnorderedMatches extends Matcher {
...
+      List<int> matched) {
+    for (int i = 0; i < matched.length; i++) {
+      if (adjacency[valueIndex][i] && !seen.contains(i)) {
```
Comment (jakemac53) [inline lib/src/iterable_matchers.dart]: Note that my suggestion above (a list of ints instead of bools for `adjacency`) would make this loop easier and more efficient also (just iterate `adjacency[valueIndex]`).

File: lib/src/iterable_matchers.dart [hunk 3]
```diff
@@ -165,30 +165,39 @@ class _UnorderedMatches extends Matcher {
...
+          if (_expected[m].matches(list[v], {})) {
+            adjacency[v][m] = true;
           }
-          ++actualPosition;
         }
-
-        if (!gotMatch) {
+      }
+      // The index into `values` matched with each matcher
+      var matched = new List<int>.filled(_expected.length, -1, growable: false);
```
Comment (jakemac53) [inline lib/src/iterable_matchers.dart]: nit: I would slightly prefer `null` to `-1`.

File: lib/src/iterable_matchers.dart [hunk 4]
```diff
@@ -206,6 +215,26 @@ class _UnorderedMatches extends Matcher {
...
+  ///
+  /// Recursively looks for new pairings whenever there is a conflict. [seen]
+  /// tracks the matchers that have already been consumed within this search.
+  bool _findPairing(List<List<bool>> adjacency, int valueIndex, Set<int> seen,
+      List<int> matched) {
+    for (int i = 0; i < matched.length; i++) {
+      if (adjacency[valueIndex][i] && !seen.contains(i)) {
+        seen.add(i);
+        if (matched[i] < 0 ||
+            _findPairing(adjacency, matched[i], seen, matched)) {
```
Comment (jakemac53) [inline lib/src/iterable_matchers.dart]: This is really hard to follow, please add a comment describing the logic or simplify it if possible.

```
If we didn't have a previous match, use this as the match and return.

Otherwise < describe logic for recursive call to _findPairing>.
```
Specifically understanding interactions of `seen` and `matched` through repeated recursive calls here is really difficult to understand.

File: test/iterable_matchers_test.dart [hunk 5]
```diff
@@ -167,14 +176,14 @@ void main() {
         unorderedMatches([3, 1]),
         "Expected: matches [<3>, <1>] unordered "
         "Actual: [1, 2] "
-        "Which: has no match for <3> at index 0");
+        "Which: has no match for <3>");
```
Comment (jakemac53) [inline test/iterable_matchers_test.dart]: fwiw, the index was actually useful here previously, especially in the case of objects that don't have good string representations. Imagine a `List<Foo>` where `Foo` has no custom `toString` method (or a not useful one), how do I know what one didn't match?

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,9 @@
+## 0.12.2
+
+* Fixed `unorderedMatches` in cases where the matchers may match more than one
```
[the remark this answers is not in the document]
Comment (natebosch) [reply CHANGELOG.md]: I could not find any breakages related to this change in a global presubmit. I scanned through the few failures and they all looked like flakes.

We do have other options though. I am fixing this because I want the correct semantics for `containsAll` - https://github.com/dart-lang/matcher/issues/72, but we could also implement the correct semantics there and deprecate this broken one.

Once we have `containsAll` then `unorderedMatches` (and by extension `unorderedEquals`) become pretty much unnecessary because these are equivalent:

```dart
expect(something, unorderedEquals(expected);

// can also be written as:
expect(something.length, expected.length);
expect(something, containsAll(expected)));
```

And `unorderedEquals` could always be substituted with `unorderedMatches` as long as none of the values were already matchers.

[the remark this answers is not in the document]
Comment (jakemac53) [reply CHANGELOG.md]: If we don't see large numbers of breakages internally then I say its safe to land :+1:

File: lib/src/iterable_matchers.dart [hunk 7]
```diff
@@ -165,30 +165,39 @@ class _UnorderedMatches extends Matcher {
...
-        var actualPosition = 0;
-        var gotMatch = false;
-        for (var actualElement in list) {
-          if (!matched[actualPosition]) {
-            if (expectedMatcher.matches(actualElement, {})) {
-              matched[actualPosition] = gotMatch = true;
-              break;
-            }
+      var adjacency = list
+          .map((_) => new List.filled(_expected.length, false, growable: false))
```
[the remark this answers is not in the document]
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: Done. In the small value case it probably makes memory usage a bit worse since I suspect the default empty list allocates more memory than these fixed lists - but I think it make the code more clear and should be an improvement for large (sparsely matching) inputs.

File: lib/src/iterable_matchers.dart (hunk 3 above)
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: Done

File: lib/src/iterable_matchers.dart [hunk 8]
```diff
@@ -165,30 +165,39 @@ class _UnorderedMatches extends Matcher {
...
+            adjacency[v][m] = true;
           }
-          ++actualPosition;
         }
-
-        if (!gotMatch) {
+      }
+      // The index into `values` matched with each matcher
+      var matched = new List<int>.filled(_expected.length, -1, growable: false);
+      for (int valueIndex = 0; valueIndex < list.length; valueIndex++) {
```
[the remark this answers is not in the document]
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: Changed to `values`.

File: lib/src/iterable_matchers.dart (hunk 1 above)
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: Done

File: lib/src/iterable_matchers.dart (hunk 2 above)
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: Yeah I like they way that looks. Also lets me pull the `seen.contains` check out of the loop into an `Iterable.where` which I think helps.

File: lib/src/iterable_matchers.dart (hunk 4 above)
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: Updated the doc comment, added some intermediate variables with descriptive names, and added a block comment at the point where we recurse.  WDUT?

File: test/iterable_matchers_test.dart (hunk 5 above)
Comment (natebosch) [reply test/iterable_matchers_test.dart]: I think I can add back the index. Do we think it's helpful to list out multiple matchers that were unmatched or should I stop at the first one like the old implementation?

File: lib/src/iterable_matchers.dart [hunk 9]
```diff
@@ -155,45 +157,53 @@ class _UnorderedMatches extends Matcher {
...
+      return 'has too many elements (${values.length} > ${_expected.length})';
+    }
 
-        if (!gotMatch) {
-          return new StringDescription()
-              .add('has no match for ')
-              .addDescriptionOf(expectedMatcher)
-              .add(' at index $expectedPosition')
-              .toString();
+    var edges = values.map((_) => <int>[]).toList(growable: false);
```
Comment (jakemac53) [inline lib/src/iterable_matchers.dart]: nit: you could use `List.filled`

File: lib/src/iterable_matchers.dart [hunk 10]
```diff
@@ -206,6 +216,31 @@ class _UnorderedMatches extends Matcher {
...
+  /// Returns [true] if the value at [valueIndex] can be paired with some
+  /// unmatched matcher and updates the state of [matched].
+  ///
+  /// If there is a conflic where multiple values may match the same matcher
+  /// recursively looks for a new place to match the old value. [seen] trackes
+  /// the matchers that have been used _during_ this search.
+  bool _findPairing(List<List<int>> edges, int valueIndex, List<int> matched,
+      [Set<int> seen]) {
+    seen ??= new Set<int>();
+    final possiblePairings = edges[valueIndex].where((m) => !seen.contains(m));
```
Comment (jakemac53) [inline lib/src/iterable_matchers.dart]: I was going to say this doesn't allow you to bail early, but then I remembered Iterables are lazy :D. Dart ftw!

File: lib/src/iterable_matchers.dart [hunk 11]
```diff
@@ -206,6 +216,31 @@ class _UnorderedMatches extends Matcher {
...
+    for (final matcherIndex in possiblePairings) {
+      seen.add(matcherIndex);
+      final previouslyMatched = matched[matcherIndex];
+      final canPlaceWithoutConflict = previouslyMatched == null;
```
Comment (jakemac53) [inline lib/src/iterable_matchers.dart]: nit: I don't think this local var is necessary, or you could eliminate `previouslyMatched` and move the null check up

File: test/iterable_matchers_test.dart (hunk 5 above)
Comment (natebosch) [reply test/iterable_matchers_test.dart]: Added back the index and shortened the rest of the message to "along with n other unmatched"

File: test/iterable_matchers_test.dart [hunk 12]
```diff
@@ -143,13 +143,22 @@ void main() {
...
   });
 
-  test('unorderedMatchess', () {
+  test('unorderedMatches', () {
     var d = [1, 2];
     shouldPass(d, unorderedMatches([2, 1]));
     shouldPass(d, unorderedMatches([greaterThan(1), greaterThan(0)]));
+    shouldPass(d, unorderedMatches([greaterThan(0), greaterThan(1)]));
+    shouldPass([2, 1], unorderedMatches([greaterThan(1), greaterThan(0)]));
+    shouldPass([2, 1], unorderedMatches([greaterThan(0), greaterThan(1)]));
```
Comment (jakemac53) [inline test/iterable_matchers_test.dart]: Due to the complexity of how this algorithm actually works, it would be good to add a more exhaustive test as well.

File: lib/src/iterable_matchers.dart (hunk 9 above)
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: I don't think I can because that would give us the same instance of a list at every index.

File: lib/src/iterable_matchers.dart (hunk 10 above)
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: 😄

File: lib/src/iterable_matchers.dart (hunk 11 above)
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: Removed `canPlaceWithoutConflict`

File: test/iterable_matchers_test.dart (hunk 12 above)
Comment (natebosch) [reply test/iterable_matchers_test.dart]: what did you have in mind? I think n > 2 and checking every ordering of expected and the values should be exhaustive?

File: lib/src/iterable_matchers.dart (hunk 9 above)
Comment (jakemac53) [reply lib/src/iterable_matchers.dart]: Sorry meant `List.generate`

File: test/iterable_matchers_test.dart (hunk 12 above)
Comment (jakemac53) [reply test/iterable_matchers_test.dart]: More items, some of which overlap with other ones, some of which are only one. Maybe something like...
```
shouldPass([1, 2, 3, 4, 5, 6], unorderedMatches([greaterThan(1), equals(2), allOf([lessThan(3), isNot(0)]), equals(0), predicate((v) => v % 2 == 1), equals(5)]));
```

Comment (jakemac53) [reply test/iterable_matchers_test.dart]: bonus points if you go through all possible orderings of those matchers and test each of them haha

Comment (natebosch) [reply test/iterable_matchers_test.dart]: Done

File: lib/src/iterable_matchers.dart (hunk 9 above)
Comment (natebosch) [reply lib/src/iterable_matchers.dart]: Sure, it's a few extra characters for now but once we can remove `new` it'll fit on one line again and I think the intent is more clear that the `.toList()` call.
